File: src/openjarvis/tools/storage/knowledge_graph.py
"""Knowledge graph storage backend — entity-relation store with pattern queries."""

from __future__ import annotations

import json
import sqlite3
import time
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any, Dict, List, Optional, Union

from openjarvis.core.config import DEFAULT_CONFIG_DIR
from openjarvis.core.registry import MemoryRegistry


@dataclass(slots=True)
class Entity:
    """A node in the knowledge graph."""
    entity_id: str
    entity_type: str   # "agent", "tool", "model", "user", "concept", etc.
    name: str
    properties: Dict[str, Any] = field(default_factory=dict)
    created_at: float = 0.0


@dataclass(slots=True)
class Relation:
    """An edge between two entities."""
    source_id: str
    target_id: str
    relation_type: str  # "used", "produced", "depends_on", "similar_to", etc.
    weight: float = 1.0
    properties: Dict[str, Any] = field(default_factory=dict)
    created_at: float = 0.0


@dataclass(slots=True)
class GraphQueryResult:
    """Result from a graph pattern query."""
    entities: List[Entity] = field(default_factory=list)
    relations: List[Relation] = field(default_factory=list)


@MemoryRegistry.register("knowledge_graph")
class KnowledgeGraphMemory:
    """SQLite-backed knowledge graph implementing MemoryBackend ABC.

    Provides standard store/retrieve/delete/clear operations plus
    graph-specific operations: add_entity(), add_relation(),
    query_pattern(), neighbors().
    """

    def __init__(
        self,
        db_path: Union[str, Path] = DEFAULT_CONFIG_DIR / "knowledge_graph.db",
        **kwargs: Any,
    ) -> None:
        self._db_path = Path(db_path)
        self._db_path.parent.mkdir(parents=True, exist_ok=True)
        self._conn = sqlite3.connect(str(self._db_path))
        self._create_tables()

    def _create_tables(self) -> None:
        self._conn.executescript("""
            CREATE TABLE IF NOT EXISTS entities (
                entity_id   TEXT PRIMARY KEY,
                entity_type TEXT NOT NULL,
                name        TEXT NOT NULL,
                properties  TEXT DEFAULT '{}',
                created_at  REAL
            );
            CREATE TABLE IF NOT EXISTS relations (
                id            INTEGER PRIMARY KEY,
                source_id     TEXT NOT NULL,
                target_id     TEXT NOT NULL,
                relation_type TEXT NOT NULL,
                weight        REAL DEFAULT 1.0,
                properties    TEXT DEFAULT '{}',
                created_at    REAL,
                FOREIGN KEY (source_id) REFERENCES entities(entity_id),
                FOREIGN KEY (target_id) REFERENCES entities(entity_id)
            );
            CREATE INDEX IF NOT EXISTS idx_relations_source ON relations(source_id);
            CREATE INDEX IF NOT EXISTS idx_relations_target ON relations(target_id);
            CREATE INDEX IF NOT EXISTS idx_relations_type ON relations(relation_type);
            CREATE INDEX IF NOT EXISTS idx_entities_type ON entities(entity_type);
        """)
        self._conn.commit()

    # -- MemoryBackend ABC --

    def store(
        self, key: str, content: str,
        metadata: Optional[Dict[str, Any]] = None,
    ) -> None:
        """Store content as an entity (MemoryBackend interface)."""
        meta = metadata or {}
        self.add_entity(Entity(
            entity_id=key,
            entity_type=meta.get("entity_type", "document"),
            name=meta.get("name", key),
            properties={"content": content, **(meta.get("properties", {}))},
        ))

    def retrieve(self, key: str) -> Optional[str]:
        """Retrieve content by entity_id (MemoryBackend interface)."""
        entity = self.get_entity(key)
        if entity:
            return entity.properties.get("content", json.dumps(entity.properties))
        return None

    def search(self, query: str, top_k: int = 5, **kwargs: Any) -> List[Dict[str, Any]]:
        """Search entities by name/type/content (MemoryBackend interface)."""
        rows = self._conn.execute(
            "SELECT entity_id, entity_type, name, properties, created_at "
            "FROM entities "
            "WHERE name LIKE ? OR entity_type LIKE ? OR properties LIKE ? "
            "LIMIT ?",
            (f"%{query}%", f"%{query}%", f"%{query}%", top_k),
        ).fetchall()
        results = []
        for row in rows:
            eid, etype, name, props_json, ts = row
            props = json.loads(props_json) if props_json else {}
            results.append({
                "key": eid,
                "content": props.get("content", ""),
                "score": 1.0,
                "metadata": {"entity_type": etype, "name": name},
            })
        return results

    def delete(self, key: str) -> bool:
        """Delete an entity and its relations."""
        self._conn.execute(
            "DELETE FROM relations"
            " WHERE source_id = ? OR target_id = ?",
            (key, key),
        )
        cur = self._conn.execute("DELETE FROM entities WHERE entity_id = ?", (key,))
        self._conn.commit()
        return cur.rowcount > 0

    def clear(self) -> None:
        """Remove all entities and relations."""
        self._conn.execute("DELETE FROM relations")
        self._conn.execute("DELETE FROM entities")
        self._conn.commit()

    # -- Graph-specific operations --

    def add_entity(self, entity: Entity) -> None:
        """Add or update an entity."""
        ts = entity.created_at or time.time()
        self._conn.execute(
            "INSERT OR REPLACE INTO entities"
            " (entity_id, entity_type, name,"
            " properties, created_at) "
            "VALUES (?, ?, ?, ?, ?)",
            (entity.entity_id, entity.entity_type, entity.name,
             json.dumps(entity.properties), ts),
        )
        self._conn.commit()

    def get_entity(self, entity_id: str) -> Optional[Entity]:
        """Get entity by ID."""
        row = self._conn.execute(
            "SELECT entity_id, entity_type, name, properties, created_at "
            "FROM entities WHERE entity_id = ?", (entity_id,),
        ).fetchone()
        if not row:
            return None
        return Entity(
            entity_id=row[0], entity_type=row[1], name=row[2],
            properties=json.loads(row[3]) if row[3] else {},
            created_at=row[4] or 0.0,
        )

    def add_relation(self, relation: Relation) -> None:
        """Add a relation between two entities."""
        ts = relation.created_at or time.time()
        self._conn.execute(
            "INSERT INTO relations"
            " (source_id, target_id, relation_type,"
            " weight, properties, created_at) "
            "VALUES (?, ?, ?, ?, ?, ?)",
            (relation.source_id, relation.target_id, relation.relation_type,
             relation.weight, json.dumps(relation.properties), ts),
        )
        self._conn.commit()

    def neighbors(
        self,
        entity_id: str,
        *,
        relation_type: Optional[str] = None,
        direction: str = "both",
        limit: int = 50,
    ) -> List[Entity]:
        """Get neighboring entities connected by relations."""
        results: List[Entity] = []

        if direction in ("out", "both"):
            sql = "SELECT target_id FROM relations WHERE source_id = ?"
            params: list = [entity_id]
            if relation_type:
                sql += " AND relation_type = ?"
                params.append(relation_type)
            sql += " LIMIT ?"
            params.append(limit)
            for row in self._conn.execute(sql, params).fetchall():
                entity = self.get_entity(row[0])
                if entity:
                    results.append(entity)

        if direction in ("in", "both"):
            sql = "SELECT source_id FROM relations WHERE target_id = ?"
            params = [entity_id]
            if relation_type:
                sql += " AND relation_type = ?"
                params.append(relation_type)
            sql += " LIMIT ?"
            params.append(limit)
            for row in self._conn.execute(sql, params).fetchall():
                entity = self.get_entity(row[0])
                if entity and entity.entity_id != entity_id:
                    results.append(entity)

        return results[:limit]

    def query_pattern(
        self,
        *,
        entity_type: Optional[str] = None,
        relation_type: Optional[str] = None,
        limit: int = 50,
    ) -> GraphQueryResult:
        """Query entities and relations matching a pattern."""
        entities: List[Entity] = []
        relations: List[Relation] = []

        if entity_type:
            rows = self._conn.execute(
                "SELECT entity_id, entity_type, name, properties, created_at "
                "FROM entities WHERE entity_type = ? LIMIT ?",
                (entity_type, limit),
            ).fetchall()
            for row in rows:
                entities.append(Entity(
                    entity_id=row[0], entity_type=row[1], name=row[2],
                    properties=json.loads(row[3]) if row[3] else {},
                    created_at=row[4] or 0.0,
                ))

        if relation_type:
            rows = self._conn.execute(
                "SELECT source_id, target_id,"
                " relation_type, weight,"
                " properties, created_at "
                "FROM relations"
                " WHERE relation_type = ? LIMIT ?",
                (relation_type, limit),
            ).fetchall()
            for row in rows:
                relations.append(Relation(
                    source_id=row[0], target_id=row[1], relation_type=row[2],
                    weight=row[3], properties=json.loads(row[4]) if row[4] else {},
                    created_at=row[5] or 0.0,
                ))

        return GraphQueryResult(entities=entities, relations=relations)

    def entity_count(self) -> int:
        row = self._conn.execute("SELECT COUNT(*) FROM entities").fetchone()
        return row[0] if row else 0

    def relation_count(self) -> int:
        row = self._conn.execute("SELECT COUNT(*) FROM relations").fetchone()
        return row[0] if row else 0

    def close(self) -> None:
        self._conn.close()


__all__ = ["Entity", "GraphQueryResult", "KnowledgeGraphMemory", "Relation"]

File: rust/crates/openjarvis-tools/src/storage/knowledge_graph.rs
//! Knowledge graph memory — SQLite entity-relation store.

use crate::storage::traits::MemoryBackend;
use openjarvis_core::{OpenJarvisError, RetrievalResult};
use parking_lot::Mutex;
use rusqlite::Connection;
use serde_json::Value;
use std::path::Path;
use uuid::Uuid;

pub struct KnowledgeGraphMemory {
    conn: Mutex<Connection>,
}

impl KnowledgeGraphMemory {
    pub fn new(db_path: &Path) -> Result<Self, OpenJarvisError> {
        if let Some(parent) = db_path.parent() {
            std::fs::create_dir_all(parent).map_err(|e| {
                OpenJarvisError::Io(std::io::Error::other(e))
            })?;
        }

        let conn = Connection::open(db_path).map_err(|e| {
            OpenJarvisError::Io(std::io::Error::other(
                e.to_string(),
            ))
        })?;

        conn.execute_batch(
            "CREATE TABLE IF NOT EXISTS entities (
                id TEXT PRIMARY KEY,
                name TEXT NOT NULL,
                entity_type TEXT DEFAULT '',
                properties TEXT DEFAULT '{}'
            );
            CREATE TABLE IF NOT EXISTS relations (
                id TEXT PRIMARY KEY,
                source_id TEXT NOT NULL,
                target_id TEXT NOT NULL,
                relation_type TEXT NOT NULL,
                properties TEXT DEFAULT '{}',
                FOREIGN KEY (source_id) REFERENCES entities(id),
                FOREIGN KEY (target_id) REFERENCES entities(id)
            );
            CREATE INDEX IF NOT EXISTS idx_relations_source ON relations(source_id);
            CREATE INDEX IF NOT EXISTS idx_relations_target ON relations(target_id);",
        )
        .map_err(|e| {
            OpenJarvisError::Io(std::io::Error::other(
                e.to_string(),
            ))
        })?;

        Ok(Self {
            conn: Mutex::new(conn),
        })
    }

    pub fn in_memory() -> Result<Self, OpenJarvisError> {
        Self::new(Path::new(":memory:"))
    }

    pub fn add_entity(
        &self,
        name: &str,
        entity_type: &str,
        properties: Option<&Value>,
    ) -> Result<String, OpenJarvisError> {
        let id = Uuid::new_v4().to_string();
        let props = properties
            .map(|p| serde_json::to_string(p).unwrap_or_default())
            .unwrap_or_else(|| "{}".to_string());

        let conn = self.conn.lock();
        conn.execute(
            "INSERT INTO entities (id, name, entity_type, properties) VALUES (?1, ?2, ?3, ?4)",
            rusqlite::params![id, name, entity_type, props],
        )
        .map_err(|e| {
            OpenJarvisError::Io(std::io::Error::other(
                e.to_string(),
            ))
        })?;

        Ok(id)
    }

    pub fn add_relation(
        &self,
        source_id: &str,
        target_id: &str,
        relation_type: &str,
        properties: Option<&Value>,
    ) -> Result<String, OpenJarvisError> {
        let id = Uuid::new_v4().to_string();
        let props = properties
            .map(|p| serde_json::to_string(p).unwrap_or_default())
            .unwrap_or_else(|| "{}".to_string());

        let conn = self.conn.lock();
        conn.execute(
            "INSERT INTO relations (id, source_id, target_id, relation_type, properties)
             VALUES (?1, ?2, ?3, ?4, ?5)",
            rusqlite::params![id, source_id, target_id, relation_type, props],
        )
        .map_err(|e| {
            OpenJarvisError::Io(std::io::Error::other(
                e.to_string(),
            ))
        })?;

        Ok(id)
    }

    pub fn neighbors(&self, entity_id: &str) -> Result<Vec<(String, String, String)>, OpenJarvisError> {
        let conn = self.conn.lock();
        let mut stmt = conn
            .prepare(
                "SELECT e.name, r.relation_type, 'outgoing'
                 FROM relations r JOIN entities e ON e.id = r.target_id
                 WHERE r.source_id = ?1
                 UNION ALL
                 SELECT e.name, r.relation_type, 'incoming'
                 FROM relations r JOIN entities e ON e.id = r.source_id
                 WHERE r.target_id = ?1",
            )
            .map_err(|e| {
                OpenJarvisError::Io(std::io::Error::other(
                    e.to_string(),
                ))
            })?;

        let results = stmt
            .query_map(rusqlite::params![entity_id], |row| {
                Ok((
                    row.get::<_, String>(0)?,
                    row.get::<_, String>(1)?,
                    row.get::<_, String>(2)?,
                ))
            })
            .map_err(|e| {
                OpenJarvisError::Io(std::io::Error::other(
                    e.to_string(),
                ))
            })?
            .filter_map(|r| r.ok())
            .collect();

        Ok(results)
    }
}

impl MemoryBackend for KnowledgeGraphMemory {
    fn backend_id(&self) -> &str {
        "knowledge_graph"
    }

    fn store(
        &self,
        content: &str,
        source: &str,
        metadata: Option<&Value>,
    ) -> Result<String, OpenJarvisError> {
        self.add_entity(content, source, metadata)
    }

    fn retrieve(
        &self,
        query: &str,
        top_k: usize,
    ) -> Result<Vec<RetrievalResult>, OpenJarvisError> {
        let conn = self.conn.lock();
        let pattern = format!("%{}%", query);
        let mut stmt = conn
            .prepare(
                "SELECT name, entity_type, properties
                 FROM entities WHERE name LIKE ?1 LIMIT ?2",
            )
            .map_err(|e| {
                OpenJarvisError::Io(std::io::Error::other(
                    e.to_string(),
                ))
            })?;

        let results = stmt
            .query_map(rusqlite::params![pattern, top_k as i64], |row| {
                Ok(RetrievalResult {
                    content: row.get::<_, String>(0)?,
                    source: row.get::<_, String>(1).unwrap_or_default(),
                    score: 1.0,
                    metadata: row
                        .get::<_, String>(2)
                        .ok()
                        .and_then(|s| serde_json::from_str(&s).ok())
                        .unwrap_or_default(),
                })
            })
            .map_err(|e| {
                OpenJarvisError::Io(std::io::Error::other(
                    e.to_string(),
                ))
            })?
            .filter_map(|r| r.ok())
            .collect();

        Ok(results)
    }

    fn delete(&self, doc_id: &str) -> Result<bool, OpenJarvisError> {
        let conn = self.conn.lock();
        let changes = conn
            .execute(
                "DELETE FROM entities WHERE id = ?1",
                rusqlite::params![doc_id],
            )
            .map_err(|e| {
                OpenJarvisError::Io(std::io::Error::other(
                    e.to_string(),
                ))
            })?;
        Ok(changes > 0)
    }

    fn clear(&self) -> Result<(), OpenJarvisError> {
        let conn = self.conn.lock();
        conn.execute_batch("DELETE FROM relations; DELETE FROM entities;")
            .map_err(|e| {
                OpenJarvisError::Io(std::io::Error::other(
                    e.to_string(),
                ))
            })?;
        Ok(())
    }

    fn count(&self) -> Result<usize, OpenJarvisError> {
        let conn = self.conn.lock();
        let count: i64 = conn
            .query_row("SELECT COUNT(*) FROM entities", [], |row| row.get(0))
            .map_err(|e| {
                OpenJarvisError::Io(std::io::Error::other(
                    e.to_string(),
                ))
            })?;
        Ok(count as usize)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_kg_entities_and_relations() {
        let kg = KnowledgeGraphMemory::in_memory().unwrap();
        let e1 = kg.add_entity("Rust", "language", None).unwrap();
        let e2 = kg.add_entity("Systems Programming", "concept", None).unwrap();
        kg.add_relation(&e1, &e2, "used_for", None).unwrap();

        let neighbors = kg.neighbors(&e1).unwrap();
        assert_eq!(neighbors.len(), 1);
        assert_eq!(neighbors[0].0, "Systems Programming");
    }
}
